acceptance/tests/environment/variables_refreshed_each_compilation.rb

```diff
@@ -0,0 +1,66 @@
+test_name 'C98115 compilation should get new values in variables on each compilation' do
+require 'puppet/acceptance/environment_utils'
+extend Puppet::Acceptance::EnvironmentUtils
+
+  app_type               = File.basename(__FILE__, '.*')
+  tmp_environment        = mk_tmp_environment_with_teardown(master, app_type)
+  fq_tmp_environmentpath = "#{environmentpath}/#{tmp_environment}"
+
+  create_remote_file(master, "#{fq_tmp_environmentpath}/environment.conf", <<-CONF)
+    environment_timeout = unlimited
+  CONF
+  # the module function loading logic is different from inside a single manifest
+  #   we exercise both here
+  on(master, "mkdir -p #{fq_tmp_environmentpath}/modules/uptime/{manifests,functions}")
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/uptime/manifests/init.pp", <<-FILE)
+    class uptime {
+      notify { 'current uptime':
+        message => uptime::my_system_uptime()
+      }
+    }
+  FILE
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/uptime/functions/my_system_uptime.pp", <<-FILE)
+    function uptime::my_system_uptime() {
+      $facts['system_uptime']
+    }
+  FILE
+  create_sitepp(master, tmp_environment, <<-SITE)
+    function bar() {
+      $facts['system_uptime']['seconds']
+    }
+    class foo::bar {
+      notify { "local_${bar()}_local": }
+    }
+    include foo::bar
+    include uptime
+  SITE
+  on(master, "chmod -R 0777 #{fq_tmp_environmentpath}/")
+
+  step "run agent in #{tmp_environment}, ensure it increments the uptime with each run" do
+    with_puppet_running_on(master,{}) do
+      uptime = nil
+      module_uptime = nil
+      local_uptime_pattern = 'local_(\d+)_local'
+      agents.each do |agent|
+        on(agent, puppet("agent -t --server #{master.hostname} --environment #{tmp_environment}"),
+           :accept_all_exit_codes => true) do |result|
+          assert_match(/Notice: #{local_uptime_pattern}/, result.stdout, 'first uptime was not as expected')
+          assert_match(/"seconds"=>\d+,/, result.stdout, 'first module uptime was not as expected')
+          uptime = result.stdout.match(/Notice: #{local_uptime_pattern}/)[0]
+          module_uptime = result.stdout.match(/"seconds"=>(\d+),/)[0]
+        end
+        sleep 1
+        on(agent, puppet("agent -t --server #{master.hostname} --environment #{tmp_environment}"),
+           :accept_all_exit_codes => true) do |result|
```

Comment:
+1 on having beaker fail based on unexpected error codes.

The asserts below do not check for desired exit codes. So, the agent run could have errors in it and pass so long as the uptime regexes are satisfied.